### PR TITLE
Show first available thumbnail in order of ZTF, SDSS, DR8, PS1 in homepage widgets

### PR DESCRIPTION
**Pull request by acrellin**
Closes https://github.com/skyportal/skyportal/issues/2020

image

### CODE REVIEW

File: skyportal/handlers/api/internal/source_views.py [hunk 1]
```diff
@@ -74,3 +78,8 @@ def post(self, obj_id):
         DBSession.add(sv)
         self.verify_and_commit()
         return self.success()
+
+
+def tIndex(t):
```
Comment (dmitryduev) [inline skyportal/handlers/api/internal/source_views.py:83]:
```suggestion
def t_index(t):
```
? I'm surprised the linter didn't complain...

File: skyportal/handlers/api/internal/source_views.py [hunk 2]
```diff
@@ -49,15 +48,20 @@ def get(self):
...
-            info = first_thumbnail_info(s.thumbnails)
             sources.append(
                 {
                     'obj_id': s.id,
                     'views': view,
                     'ra': s.ra,
                     'dec': s.dec,
-                    'public_url': info[0],
-                    'is_grayscale': info[1],
+                    'thumbnails': [
```
Comment (dmitryduev) [inline skyportal/handlers/api/internal/source_views.py:57]: 👍🏻

**Review by dmitryduev — APPROVED** [review]
LGTM @acrellin! Just one minor comment.

File: skyportal/handlers/api/internal/recent_sources.py [hunk 3]
```diff
@@ -4,6 +4,7 @@
 from baselayer.app.access import auth_or_token
 from ...base import BaseHandler
 from ....models import Obj, Source
+from .source_views import tIndex
```
Comment (stefanv) [inline skyportal/handlers/api/internal/recent_sources.py:7]: Camelcase in Python land!  Should become t_index

File: static/js/components/RecentSources.jsx [hunk 4]
```diff
@@ -137,7 +137,31 @@ const RecentSourcesList = ({ sources, styles }) => {
...
+          // Use first thumbnail available
+          const fetchThumbnail = async (index) => {
+            const response = await fetch(source.thumbnails[index]?.public_url, {
+              mode: "no-cors",
+            });
+            return response.status;
+          };
+          let idx = 0;
+          let statusCode = 0;
+          while (idx < source.thumbnails.length - 1) {
```
Comment (stefanv) [inline static/js/components/RecentSources.jsx:149]: In general, while loops are dangerous and overly verbose and should be replaced by for-loops.

In JavaScript, this loop can be replaced by [`.each`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/every), for example.  Alternatively, you can use a for loop and break out of it.

File: static/js/components/TopSources.jsx [hunk 5]
```diff
@@ -114,7 +114,31 @@ const TopSourcesList = ({ sources, styles }) => {
...
+          let idx = 0;
+          let statusCode = 0;
+          while (idx < source.thumbnails.length - 1) {
+            if (
+              source.thumbnails[idx]?.public_url &&
+              source.thumbnails[idx].public_url.startsWith("/static")
+            ) {
+              break;
+            }
+            statusCode = fetchThumbnail(idx);
```
Comment (stefanv) [inline static/js/components/TopSources.jsx:133]: Shoudn't an await be used here?

**Review by stefanv — COMMENTED** [review]
Looks good, thanks Ari.  I have some small requests.

File: skyportal/handlers/api/internal/recent_sources.py (hunk 3 above)
Comment (acrellin) [reply skyportal/handlers/api/internal/recent_sources.py:7]: Yes, agreed (I didn't author this, just moved it around, as you can see from the diff -- will update anyways).

Comment (stefanv) [reply skyportal/handlers/api/internal/recent_sources.py:7]: 👍

File: static/js/components/RecentSources.jsx (hunk 4 above)
Comment (acrellin) [reply static/js/components/RecentSources.jsx:149]: I don't think it's possible to break out of a [`forEach`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach) loop (I assume that's what you meant). I can replace it with a for loop if you'd like.

Comment (stefanv) [reply static/js/components/RecentSources.jsx:149]: I meant a `.each` loop---link in comment above.

Comment (acrellin) [reply static/js/components/RecentSources.jsx:149]: That links to `.every`

Comment (acrellin) [reply static/js/components/RecentSources.jsx:149]: Which doesn't really apply to this use case

Comment (stefanv) [reply static/js/components/RecentSources.jsx:149]: You can use `.some` or `.every`.

Comment (stefanv) [reply static/js/components/RecentSources.jsx:149]:
```
source.thumbnails.some((thumb) =>
  ...
  return statusCode == 200;
);
```

**Comment by profjsb** [comment]
@stefanv @acrellin is this PR still relevant? If so, let's get it merged.

**Comment by dmitryduev** [comment]
@profjsb it is relevant, but as I recall @acrellin wanted to finish something here.

**Comment by acrellin** [comment]
@stefanv ready for review.  I switched to simply incrementing the index used to key into `source.thumbnails` if a given index results in an error when fetching the image.  The server returns the thumbnail list in the desired order.

**Comment by acrellin** [comment]
@stefanv ready for review when you have a moment.

**Comment by stefanv** [comment]
Let's give it a try!

**Pull request merged**